File: runtime/go/prompty/jinjasubset/tokenizer.go
package jinjasubset

import (
	"fmt"
	"strings"
	"unicode"
)

type tokenType int

const (
	tokenText tokenType = iota
	tokenExpr
	tokenStmt
	tokenComment
)

type token struct {
	typ       tokenType
	value     string
	trimLeft  bool
	trimRight bool
}

func tokenize(template string) ([]token, error) {
	raw := []token{}
	i, textStart := 0, 0
	for i < len(template) {
		if i+2 <= len(template) {
			two := template[i : i+2]
			typ, close, ok := opener(two)
			if ok {
				if i > textStart {
					raw = append(raw, token{typ: tokenText, value: template[textStart:i]})
				}
				closeIdx := strings.Index(template[i+2:], close)
				if closeIdx < 0 {
					return nil, &TemplateSyntaxError{Message: fmt.Sprintf("unclosed %q tag at offset %d", two, i)}
				}
				closeIdx += i + 2
				inner := template[i+2 : closeIdx]
				trimLeft := strings.HasPrefix(inner, "-")
				trimRight := strings.HasSuffix(inner, "-")
				if trimLeft {
					inner = inner[1:]
				}
				if trimRight {
					inner = inner[:len(inner)-1]
				}
				if typ == tokenComment {
					raw = append(raw, token{typ: tokenComment, trimLeft: trimLeft, trimRight: trimRight})
				} else {
					raw = append(raw, token{typ: typ, value: strings.TrimSpace(inner), trimLeft: trimLeft, trimRight: trimRight})
				}
				i = closeIdx + len(close)
				textStart = i
				continue
			}
		}
		i++
	}
	if textStart < len(template) {
		raw = append(raw, token{typ: tokenText, value: template[textStart:]})
	}
	applyTrims(raw)
	out := make([]token, 0, len(raw))
	for _, tok := range raw {
		if tok.typ != tokenComment {
			out = append(out, tok)
		}
	}
	return out, nil
}

func opener(two string) (tokenType, string, bool) {
	switch two {
	case "{{":
		return tokenExpr, "}}", true
	case "{%":
		return tokenStmt, "%}", true
	case "{#":
		return tokenComment, "#}", true
	default:
		return tokenText, "", false
	}
}

func applyTrims(tokens []token) {
	for i := range tokens {
		if tokens[i].typ == tokenText {
			continue
		}
		if tokens[i].trimLeft && i > 0 && tokens[i-1].typ == tokenText {
			tokens[i-1].value = strings.TrimRightFunc(tokens[i-1].value, unicode.IsSpace)
		}
		if tokens[i].trimRight && i+1 < len(tokens) && tokens[i+1].typ == tokenText {
			tokens[i+1].value = strings.TrimLeftFunc(tokens[i+1].value, unicode.IsSpace)
		}
	}
}

File: runtime/rust/prompty/src/jinja_subset/tokenizer.rs
//! Lexical scanner: splits a raw template into literal text and tag regions
//! (`{{ }}` expressions, `{% %}` statements, `{# #}` comments), applying the
//! `{%- … -%}` / `{{- … -}}` whitespace-trim markers. Ported from the reference
//! tokenizer; comments carry trim semantics but produce no node.

use super::RenderError;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum TokenType {
    Text,
    Expr,
    Stmt,
    Comment,
}

#[derive(Debug, Clone)]
pub(crate) struct Token {
    pub ty: TokenType,
    pub value: String,
    pub trim_left: bool,
    pub trim_right: bool,
}

/// Find `needle` in `haystack` at or after `start`, returning the char index.
fn find_sub(haystack: &[char], needle: &[char], start: usize) -> Option<usize> {
    if needle.is_empty() || start + needle.len() > haystack.len() {
        return None;
    }
    let last = haystack.len() - needle.len();
    (start..=last).find(|&i| &haystack[i..i + needle.len()] == needle)
}

/// Tokenize `template` into a flat stream of text and tag tokens, with comment
/// tokens removed after trim application.
pub(crate) fn tokenize(template: &str) -> Result<Vec<Token>, RenderError> {
    // Work over chars so multi-byte content never splits a UTF-8 boundary; the
    // delimiters themselves are ASCII.
    let chars: Vec<char> = template.chars().collect();
    let n = chars.len();
    let openers: [(&str, TokenType, &str); 3] = [
        ("{{", TokenType::Expr, "}}"),
        ("{%", TokenType::Stmt, "%}"),
        ("{#", TokenType::Comment, "#}"),
    ];

    let mut raw: Vec<Token> = Vec::new();
    let mut i = 0usize;
    let mut text_start = 0usize;

    while i < n {
        let two: String = if i + 2 <= n {
            chars[i..i + 2].iter().collect()
        } else {
            String::new()
        };

        if let Some(&(open, kind, close)) = openers.iter().find(|(o, _, _)| *o == two) {
            if i > text_start {
                raw.push(Token {
                    ty: TokenType::Text,
                    value: chars[text_start..i].iter().collect(),
                    trim_left: false,
                    trim_right: false,
                });
            }

            let close_chars: Vec<char> = close.chars().collect();
            let close_idx = match find_sub(&chars, &close_chars, i + 2) {
                Some(idx) => idx,
                None => {
                    return Err(RenderError::Syntax(format!(
                        "Unclosed '{open}' tag at offset {i}"
                    )));
                }
            };

            let mut inner: String = chars[i + 2..close_idx].iter().collect();
            let trim_left = inner.starts_with('-');
            let trim_right = inner.ends_with('-');
            if trim_left {
                inner = inner[1..].to_string();
            }
            if trim_right {
                inner = inner[..inner.len() - 1].to_string();
            }

            if kind == TokenType::Comment {
                // No node, but the trim markers still apply to neighbours.
                raw.push(Token {
                    ty: TokenType::Comment,
                    value: String::new(),
                    trim_left,
                    trim_right,
                });
            } else {
                raw.push(Token {
                    ty: kind,
                    value: inner.trim().to_string(),
                    trim_left,
                    trim_right,
                });
            }

            i = close_idx + close_chars.len();
            text_start = i;
        } else {
            i += 1;
        }
    }

    if text_start < n {
        raw.push(Token {
            ty: TokenType::Text,
            value: chars[text_start..].iter().collect(),
            trim_left: false,
            trim_right: false,
        });
    }

    apply_trims(&mut raw);
    Ok(raw
        .into_iter()
        .filter(|t| t.ty != TokenType::Comment)
        .collect())
}

/// Apply `{%- -%}`/`{{- -}}` trim markers: a tag with `trim_left` trims the
/// trailing whitespace of the preceding text token; `trim_right` trims the
/// leading whitespace of the following text token.
fn apply_trims(tokens: &mut [Token]) {
    for idx in 0..tokens.len() {
        if tokens[idx].ty == TokenType::Text {
            continue;
        }
        let (trim_left, trim_right) = (tokens[idx].trim_left, tokens[idx].trim_right);
        if trim_left && idx > 0 && tokens[idx - 1].ty == TokenType::Text {
            let trimmed = tokens[idx - 1].value.trim_end().to_string();
            tokens[idx - 1].value = trimmed;
        }
        if trim_right && idx + 1 < tokens.len() && tokens[idx + 1].ty == TokenType::Text {
            let trimmed = tokens[idx + 1].value.trim_start().to_string();
            tokens[idx + 1].value = trimmed;
        }
    }
}
